packages/headless/src/controllers/triggers/headless-query-trigger.ts

```diff
@@ -0,0 +1,87 @@
+import {SearchEngine} from '../../app/search-engine/search-engine';
+import {ConfigurationSection, TriggerSection} from '../../state/state-sections';
+import {configuration, triggers} from '../../app/reducers';
+import {buildController, Controller} from '../controller/headless-controller';
+import {loadReducerError} from '../../utils/errors';
+import {executeSearch} from '../../features/search/search-actions';
+import {
+  updateQuery,
+  UpdateQueryActionCreatorPayload,
+} from '../../features/query/query-actions';
+import {logQueryTrigger} from '../../features/triggers/trigger-analytics-actions';
+
+/**
+ * The `QueryTrigger` controller handles query triggers.
+ */
+export interface QueryTrigger extends Controller {
+  /**
+   * the state of the `QueryTrigger` controller.
+   */
+  state: QueryTriggerState;
+}
+
+/**
+ * A scoped and simplified part of the headless state that is relevant to the `QueryTrigger` controller.
+ */
+export interface QueryTriggerState {
+  /**
+   * The new query to perform a search with after receiving a query trigger.
+   */
+  newQuery: string | null;
+
+  /**
+   * The query used to perform the search that received a query trigger in its response.
+   */
+  prevQuery: string | undefined;
+}
+
+/**
+ * Creates a `QueryTrigger` controller instance.
+ *
+ * @param engine - The headless engine.
+ * @returns A `QueryTrigger` controller instance.
+ * */
+export function buildQueryTrigger(engine: SearchEngine): QueryTrigger {
+  if (!loadQueryTriggerReducers(engine)) {
+    throw loadReducerError;
+  }
+
+  const controller = buildController(engine);
+  const {dispatch} = engine;
+
+  const getState = () => engine.state;
+
+  return {
+    ...controller,
+
+    subscribe(listener: () => void) {
+      const strictListener = () => {
+        if (getState().triggers.query) {
+          listener();
+          const updateQueryPayload: UpdateQueryActionCreatorPayload = {
+            q: getState().triggers.query!,
+          };
+          dispatch(updateQuery(updateQueryPayload));
+          dispatch(executeSearch(logQueryTrigger()));
+          getState().triggers.query = '';
+        }
```

Comment:
I find it strange to only handle query triggers when we subscribe to the controller. Could we handle them when the controller is built instead?

Reply:
If we only do it when the controller is built then that would require for the user to check for query triggers by himself in every response and then build a new query trigger controller every time to handle it. Whereas with subscribe, from what I understand, the controller is initialized once and on every engine state change, it will check the engine's state and see if there are any query triggers it must handle.

Reply:
Yes exactly! When you subscribe, you set a listener and perhaps want to do something else when the trigger is executed e.g. display a message

Reply:
could a subscribe be called by the controller itself when it's initialized? Perhaps I'm misunderstanding the purpose of the controller.

Ideally, I would avoid having behaviour that happens only after we add a listener, since this isn't something I've commonly seen in any other framework or library.

Reply:
I think the standard way for users to use the headless controllers is to initialize them and then subscribe by themselves. We give them the freedom and responsibility of subscribing on their own because this allows them to pass in a function as a parameter to the subscribe method to be called whenever the state of the engine changes.

Reply:
Our initial designs had a hook in the options, but that was too different of a pattern, so we used the subscribe method

Reply:
Thinking about this more @btaillon  has a point. For the previous redirection trigger, it implied a required action on the part of the user, since they *should* modify their Url. The subscribe enforcement was a good idea here.

For this component, the user doesn't really need to do anything, we trigger the query and log analytics for him. He can still subscribe to the component to see the trigger info change.

We should discuss this more, but I think we could subscribe to the engine on registration.

Thoughts?

Reply:
Discussed and resolved in scrum